### DISCUSSION TITLE
CO2-Emissionsfaktor sollte konfigurierbar sein

### DISCUSSION THREAD

**Discussion by boecks** [solectrus/solectrus#3181] [Ideas]
In Österreich werden dreierlei Emissionsfaktoren auf der Homepage des Umweltbundesamtes für Strom angegeben.
Siehe: https://secure.umweltbundesamt.at/co2mon/co2mon.html

Je nachdem, welcher Strommix auf den eigenen Haushalt zutrifft, ist ein anderer Faktor anwendbar:

**Stromaufbringung   Österreich: 0,23kg/kWh**
_Bei   der  Stromaufbringung Österreich werden   die inländische Stromerzeugung und die Stromimporte berücksichtigt. Der   Emissionsfaktor für die inländische Stromerzeugung wird anhand des   Kraftwerkparks ermittelt. Für die Stromimporte werden die Emissionsfaktoren   der Stromerzeugung der Importländer herangezogen._

**Kraftwerkspark Österreich: 0,17kg/kWh**
_Es   wird der österreichische Kraftwerkspark d.h. sämtliche Stromerzeugungsanlagen   auf österreichischen Staatsgebiet berücksichtigt. Somit wird wie bei der   österreichischen Luftschadstoffinventur das Territorialprinzip angewendet._

**Umweltzeichen "Grüner   Strom": 0,01kg/kWh**
_Das   österreichische Umweltzeichen definiert in der Version 5.0 vom Jänner 2018   die Kriterien für die Stromerzeugung. Der Strom muss zur Gänze aus   erneuerbaren Energieträgern stammen und bestimmte Kriterien erfüllen. Diese   Kriterien sind in der Richtlinie "UZ 46 - Grüner Strom" definiert.   Der Emissionsfaktor wird basierend auf dem Stromerzeugungsmix der größten   Anbieter ermittelt._

Gemäß Prinzip: andere Länder andere Sitten, ist das vermutlich auch für andere Länder als DE und AT zutreffend.

Ich schlage vor, entweder den Faktor aus dem ENV zu fetchen oder
den Emissionsfaktor analog den Strompreisen im GUI konfigurierbar zu machen.

**Comment by ledermann** [comment]
Interessanter Einblick, vielen Dank! 

Eine ENV-Variable einzuführen, wäre schnell gemacht. Was meiner Ansicht nach aber nicht berücksichtigt wird, ist der zeitliche Aspekt. Die derzeitig hartkodierten 401 g/kWh stammen ja nicht nur aus Deutschland, sondern gelten auch nur für 2019. 

Um es korrekt zu machen, müsste man also den Emissionsfaktor jährlich erfassbar machen. Also vergleichbar mit den Strompreisen. Denn auch wenn ich in zwei Jahren den exakt selben PV-Strom erzeugt habe, sinkt meine CO-Einsparung, wenn der Strommix in meinem Land grüner geworden ist.

Insofern stellt sich die Frage: Wollen wir das wirklich genau berechnen (und somit die Faktoren jahresbezogen per UI komfortabel erfassbar machen) oder genügt die grobe Methode (ein Faktor per ENV-Datei)?

**Comment by spirit0607** [comment]
Wie wäre es denn mit der Möglichkeit "Electricity Maps" einzubinden? 

Diese stellt eine kostenfreie API mit dem aktuellen CO2 Mix für den privaten Gebrauch zur Verfügung.

https://www.electricitymaps.com/free-tier-api

Somit hätte man sogar einen minütlich korrekten Wert und eine wirklich genau Auskunft darüber, wie hoch die CO2 Ersparnis tatsächlich ist. Ich stelle mir die Integration in Solectrus analog dem WeatherForecast vor. 

Wenn sich jemand nicht registrieren möchte, oder den Dienst nicht nutzen mag, könnte ja trotzdem per env einen fixen Wert eingeben. 

Vll. ist das ja umsetzbar.

**Comment by ledermann** [comment]
Das Angebot von @electricitymaps kannte ich bislang gar nicht, sehr cool, danke für den Hinweis, @spirit0607! Ich klicke mich gerade durch https://app.electricitymaps.com/zone/DE. Den bisher von SOLECTRUS verwendeten Emissionsfaktor von 401 g/kWh für DE finde ich dort zwar nicht (auch nicht bei 2019, wo er eigentlich herkommt), aber die Größenordnung stimmt. Ähnliches gilt für die oben von @AlpenFlizzer genannten Werte für AT.

Der korrekte API-Endpunkt wäre wohl dieser hier:
https://api.electricitymap.org/v3/carbon-intensity/latest?zone=DE

Dieser Wert könnte von einem neuen (noch zu erstellendem) Collector kontinuierlich abgefragt und in InfluxDB gespeichert werden. Bei der Darstellung in SOLECTRUS könnte er dann mit der PV-Erzeugung multipliziert werden. Das würde viel mehr Sinn ergeben als der jetzige konstante Faktor. 

Und ja, Fallback auf ENV-Variable wäre nützlich. Und da es eine [Integration für Home Assistant](https://www.home-assistant.io/integrations/co2signal/) gibt, könnte SOLECTRUS den Wert prinzipiell auch dort abholen, über den MQTT-Collector.

Historische Daten wären ideal, seit Inbetriebnahme der eigenen PV-Anlage. Aber die sind [im Free-Plan leider nicht enthalten](https://www.electricitymaps.com/pricing).

Insgesamt eine spannende Sache, wie sehen das die anderen?

**Comment by boecks** [comment]
Für meinen Use Case und Anspruch wäre eine Integration in den MQTT Collector das Mittel der Wahl. Bei mir läuft im Hintergrund auch noch Home Assistant und EVCC - beide können die API bereits konsumieren.
EVCC darüber hinaus auch noch Grünstromindex für DE und National Grid ESO für UK, siehe https://docs.evcc.io/docs/devices/tariffs#co-vorhersage

Habe es auch mal schnell in EVCC für den Grünstromindex versucht. Über den MQTT Broker ist dann auch folgendes Topic befüllt:
evcc/site/tariffCo2 zB um 17:10 Uhr am 13.06.2024 tariffCo2=44 für PLZ 98708 in DE

Leider konsumiert EVCC von Electricitymap nicht den Wert des `latest `endpoint, sonder des `forecast `endpoint und dafür braucht man auch einen bezahlten Token. Also wäre an der Stelle Home Assistant die einfachere Wahl, der dann via MQTT publisht. (Auch gerade getestet, ist innerhalb 5 Minuten eingerichtet...)

Ein extra Collector macht Solectrus zwar von unabhängiger von umgebenden Services (HA, EVCC,...) aber erscheint aufgrund des soliden MQTT Collectors nicht unbedingt erforderlich.

PS: Coole Sache, dass @spirit0607 diese Services auf den Tisch gebracht hat. Ist natürlich viel sinnvoller als statische Werte...

**Comment by StevieC121176** [comment]
Bei mir wäre es ähnlich wie bei @AlpenFlizzer . Ich nutze auch EVCC und das stellt bei mir den Stromindex über "Grünstromindex" bereit. Per MQTT Collector könnte der Wert dann in Solectrus genutzt werden.

**Comment by spirit0607** [comment]
> Ein extra Collector macht Solectrus zwar von unabhängiger von umgebenden Services (HA, EVCC,...) aber erscheint aufgrund des soliden MQTT Collectors nicht unbedingt erforderlich

Das sehe ich auch so. Ich nutze auch sowohl evcc als auch Home Assistant. Jedoch kommt es ja auch auf die zukünftige generelle Auslegung von Solectrus an. Bei einem integrierten Collector macht man sich schon unabhängiger und für neue User ggf. leichter konfigurierbarer, wenn die z.B die Eingabe eines Tokes o.ä. beispielsweise über eine GUI funktioniert. Auf der anderen Seite werden die meisten User, welche keine SENEC Anlage haben, oder die ein weiteres externes Gerät betreiben, wie z.B. eine Wärmepumpe eh MQTT und ein weiters System wie Home Assistant oder evcc benötigen.

**Comment by ledermann** [comment]
Danke für die Rückmeldungen! Auch ich habe auf die Schnelle den HomeAssistant so einrichten können, dass die API von electricitymaps.com stündlich angezapft wird. Dennoch denke ich, dass ein Collector notwendig wäre, da viele SOLECTRUS-User gar keinen HomeAssistant einsetzen. 

Ich habe das Thema mal in eine Diskussion abgeändert, weil noch ein paar inhaltliche Dinge geklärt werden sollten, bevor etwas umgesetzt wird. Also zusätzlich zur Frage, wo genau der Faktor herkommt. 

Bei der Betrachtung des zeitlichen Verlaufs des Emissionsfaktors wird deutlich, dass dieser zumindest in DE nachts deutlich ansteigt und tagsüber abfällt. Klar, nachts entfällt die Photovoltaik und es müssen vermehrt fossile Stromerzeuger ran. Aber nachts habe ich ja den Stromspeicher. Daher stellt sich mir die Frage, wie der Emissionsfaktor korrekt eingesetzt werden muss, um die persönliche CO₂-Reduzierung zu ermitteln. Denn die einfache Formel `PV-Erzeugung * Emissionsfaktor` erscheint mir nicht richtig. Ich meine, man müsste es so rechnen:

Für jede Stunde: `(Stromverbrauch - Netzbezug + Einspeisung) * Emissionsfaktor`, wobei `Stromverbrauch := Haus + Wallbox + Wärmepumpe`.

Denn: Für alles, was ich selbst verbrauche, aber nicht aus dem Netz ziehe, reduziere ich selbst das emittierte CO₂ in meiner Zone. Und zusätzlich sorge ich mit der Einspeisung dafür, dass der Stromverbrauch meiner Nachbarn weniger CO₂ verursacht.

Wendet man diese Formel auf einen größeren Zeitraum an und verwendet einen konstanten Faktor, ändert sich natürlich kaum etwas zur einfachen Formel nur mit der PV-Erzeugung. Aber da der Emissionsfaktor schwankt und die Berechnung stundenweise erfolgen muss, wird das Ergebnis doch ein anderes sein.

Naja, das bringt mir zur Frage, was es bringt, wenn man das eingesparte CO₂ mit anderen PV-Betreibern vergleichen will, aber jede App ganz anders rechnet.

**Comment by spirit0607** [comment]
Also meiner Meinung nach müsste man auf jeden Fall den Akku mit in die Berechnung einfließen lassen. Wenn ich z.B. in der Nacht eine kWh aus dem Akku ziehe und nicht aus dem Netz habe ich mit dieser kWh vermutlich eine höhere CO2 Ersparnis, als über den Direktverbrauch am Tag (eben genau aus dem Grund, dass nachts die PV Produktion entfällt und vermehrt fossile Stromerzeuger ran müssen). Im Gegenzug muss aber dann natürlich auch die Ladeleistung berücksichtig werden. Ich habe gerade mal ChatGPT um ein Beispiel gebeten. Hier das Resultat als "Denkanstoß" und Diskusionsgrundlage ;-)



` Formel:
\[ \text{CO₂-Reduktion} = \sum_{t=1}^{T} \left( (\text{SV}_t - \text{NB}_t + \text{E}_t - \text{Ladung}_t + \text{Entladung}_t) \times \text{EF}_t \right) \]

Dabei ist:
- \( \text{SV}_t \) der Stromverbrauch in Stunde \( t \).
- \( \text{NB}_t \) der Netzbezug in Stunde \( t \).
- \( \text{E}_t \) die Einspeisung in Stunde \( t \).
- \( \text{Ladung}_t \) die Strommenge, die in Stunde \( t \) in den Batteriespeicher geladen wird.
- \( \text{Entladung}_t \) die Strommenge, die in Stunde \( t \) aus dem Batteriespeicher entladen wird.
- \( \text{EF}_t \) der Emissionsfaktor in Stunde \( t \).

### Beispiel-Daten (mit Batteriespeicher):

| Stunde | SV (kWh) | NB (kWh) | E (kWh) | Ladung (kWh) | Entladung (kWh) | EF (kg CO₂/kWh) |
|--------|----------|----------|---------|--------------|-----------------|-----------------|
| 1      | 5        | 3        | 1       | 1            | 0               | 0.6             |
| 2      | 4        | 2        | 0       | 2            | 0               | 0.5             |
| 3      | 3        | 1        | 0       | 2            | 0               | 0.4             |
| 4      | 6        | 2        | 0       | 0            | 2               | 0.7             |
| 5      | 5        | 2        | 1       | 0            | 1               | 0.6             |
| 6      | 4        | 1        | 1       | 0            | 1               | 0.5             |
| 7      | 3        | 0        | 1       | 0            | 2               | 0.4             |
| 8      | 6        | 4        | 0       | 0            | 0               | 0.7             |

### Berechnung pro Stunde:

#### Stunde 1:
\[ \text{CO₂-Reduktion}_1 = (\text{SV}_1 - \text{NB}_1 + \text{E}_1 - \text{Ladung}_1 + \text{Entladung}_1) \times \text{EF}_1 \]
\[ = (5 - 3 + 1 - 1 + 0) \times 0.6 \]
\[ = 2 \times 0.6 \]
\[ = 1.2 \text{ kg CO₂} \]

#### Stunde 2:
\[ \text{CO₂-Reduktion}_2 = (\text{SV}_2 - \text{NB}_2 + \text{E}_2 - \text{Ladung}_2 + \text{Entladung}_2) \times \text{EF}_2 \]
\[ = (4 - 2 + 0 - 2 + 0) \times 0.5 \]
\[ = 0 \times 0.5 \]
\[ = 0 \text{ kg CO₂} \]

#### Stunde 3:
\[ \text{CO₂-Reduktion}_3 = (\text{SV}_3 - \text{NB}_3 + \text{E}_3 - \text{Ladung}_3 + \text{Entladung}_3) \times \text{EF}_3 \]
\[ = (3 - 1 + 0 - 2 + 0) \times 0.4 \]
\[ = 0 \times 0.4 \]
\[ = 0 \text{ kg CO₂} \]

#### Stunde 4:
\[ \text{CO₂-Reduktion}_4 = (\text{SV}_4 - \text{NB}_4 + \text{E}_4 - \text{Ladung}_4 + \text{Entladung}_4) \times \text{EF}_4 \]
\[ = (6 - 2 + 0 - 0 + 2) \times 0.7 \]
\[ = 6 \times 0.7 \]
\[ = 4.2 \text{ kg CO₂} \]

#### Stunde 5:
\[ \text{CO₂-Reduktion}_5 = (\text{SV}_5 - \text{NB}_5 + \text{E}_5 - \text{Ladung}_5 + \text{Entladung}_5) \times \text{EF}_5 \]
\[ = (5 - 2 + 1 - 0 + 1) \times 0.6 \]
\[ = 5 \times 0.6 \]
\[ = 3.0 \text{ kg CO₂} \]

#### Stunde 6:
\[ \text{CO₂-Reduktion}_6 = (\text{SV}_6 - \text{NB}_6 + \text{E}_6 - \text{Ladung}_6 + \text{Entladung}_6) \times \text{EF}_6 \]
\[ = (4 - 1 + 1 - 0 + 1) \times 0.5 \]
\[ = 5 \times 0.5 \]
\[ = 2.5 \text{ kg CO₂} \]

#### Stunde 7:
\[ \text{CO₂-Reduktion}_7 = (\text{SV}_7 - \text{NB}_7 + \text{E}_7 - \text{Ladung}_7 + \text{Entladung}_7) \times \text{EF}_7 \]
\[ = (3 - 0 + 1 - 0 + 2) \times 0.4 \]
\[ = 6 \times 0.4 \]
\[ = 2.4 \text{ kg CO₂} \]

#### Stunde 8:
\[ \text{CO₂-Reduktion}_8 = (\text{SV}_8 - \text{NB}_8 + \text{E}_8 - \text{Ladung}_8 + \text{Entladung}_8) \times \text{EF}_8 \]
\[ = (6 - 4 + 0 - 0 + 0) \times 0.7 \]
\[ = 2 \times 0.7 \]
\[ = 1.4 \text{ kg CO₂} \]

### Gesamt-CO₂-Reduktion:
\[ \text{CO₂-Reduktion}_{\text{gesamt}} = \sum_{t=1}^{8} \text{CO₂-Reduktion}_t \]
\[ = 1.2 + 0 + 0 + 4.2 + 3.0 + 2.5 + 2.4 + 1.4 \]
\[ = 14.7 \text{ kg CO₂} \]

In diesem Beispiel ergibt sich eine Gesamt-CO₂-Reduktion von 14.7 kg CO₂ über den betrachteten Zeitraum von 8 Stunden, unter Berücksichtigung des Einflusses des Batteriespeichers.`

**Comment by spirit0607** [comment]
<img width="672" alt="Bildschirmfoto 2024-06-13 um 22 57 18" src="https://github.com/solectrus/solectrus/assets/69988146/c3a2dfec-c175-4845-8e44-465dbf49d564">
<img width="672" alt="Bildschirmfoto 2024-06-13 um 22 57 50" src="https://github.com/solectrus/solectrus/assets/69988146/bed17ef0-7f92-4199-9f23-a792725c84b4">
<img width="729" alt="Bildschirmfoto 2024-06-13 um 22 58 07" src="https://github.com/solectrus/solectrus/assets/69988146/a7218ae6-9b4e-4ede-9dd6-92b389c1039f">


Irgendwie klappt das mit der Formatierung der Formeln nicht. Daher nochmal als Screenshot

**Comment by ledermann** [comment]
Ich fasse kurz zusammen:

- Mein Vorschlag: `(Stromverbrauch - Netzbezug + Einspeisung) * Emissionsfaktor`
- ChatGPT: `(Stromverbrauch - Netzbezug + Einspeisung - Akkubeladung + Akkuentladung) * Emissionsfaktor`

Ich meine, dass ChatGPT falsch liegt, weil die Wirkung des Akkus implizit in Netzbezug und Einspeisung bereits enthalten ist: **Weil** ich Strom aus dem Akku entnehme, ist der Netzbezug kleiner. Und **weil** ich Strom in den Akku lade, ist die Einspeisung geringer.

Entscheidend ist ja, dass durch meine PV-Anlage mit Speicher insgesamt weniger Strom aus dem Netz bezogen wird (entweder durch mich selbst oder durch meine Nachbarn, die von meiner Einspeisung profitieren). Nur der Netzbezug hat Einfluss auf die CO₂-Emission. 

Hier ein (exotisches) Zahlenbeispiel zur Verdeutlichung.

Sonne scheint, alle Verbraucher sind ausgeschaltet, Akku wird geladen:
- PV-Erzeugung: 2 kWh
- Akkubeladung: 2 kWh
- alles andere: 0 kWh

Mit meiner Formel wäre die Reduktion an CO₂ gleich `0`. Korrekt, wie ich finde. Denn dadurch, dass ich Sonnenstrom in meinen Akku speichere, wird kein deutsches Kohlekraftwerk entlastet. Mit der Formel von ChatGPT wäre die Ersparnis aber `- 2 kWh * Emissionsfaktor`, also negativ (!) und somit falsch.

Fortführung des Beispiels, eine Stunde später:

Sonne scheint nicht mehr, ich schalte meine Verbraucher ein, Akku wird entladen:
- Stromverbrauch: 2 kWh
- Akkuentladung: 2 kWh
- alles andere: 0 kWh

Mit meiner Formel wäre die Reduzierung des CO₂: `2 kWh * Emissionsfaktor`. Das ist korrekt, dann jetzt entlaste ich wirklich die Kohlekraftwerke - und zwar in Höhe meines Verbrauches, da ich mich selbst aus dem Akku versorge. Mit der Formel von ChatGPT wäre die Ersparnis aber doppelt so hoch, nämlich `4 kWh * Emissionsfaktor`, also falsch.

**Comment by boecks** [comment]
Wahrscheinlich ist meine Aussage unqualifiziert, aber ist es nicht letzendlich so, dass ich den aktuellen CO2 Faktor schlicht auf die bezogene Energie anwende und um die Einsparung zu erhalten dann den Hausverbrauch (gesamt) fiktiv mit dem Faktor multipliziert gegenüberstelle?
Wenn man die Akkubeladung Entladung rechnet, dann rechnet man diesen Strom ja doppelt, weil produziert wird er ja nur einmal?

Also für einen Beispieltag:
```
CO2_Faktor: 2g/kWh
Netzbezug: 5kWh
PV Produktion: 10kWh
CO2_Netz: 5*2=10g
CO2_PV: 10*2=20g (fiktiv)
CO2_Einsparung=10g-20g=-10g (CO2_Netz-CO2_PV)
```
Nach obiger rechnung habe ich also 10g eingespart. Da ist irrelevant, ob Akku oder nicht.

**Comment by spirit0607** [comment]
> Sonne scheint, alle Verbraucher sind ausgeschaltet, Akku wird geladen:
> 
> PV-Erzeugung: 2 kWh
> Akkubeladung: 2 kWh
> alles andere: 0 kWh
> Mit meiner Formel wäre die Reduktion an CO₂ gleich 0. Korrekt, wie ich finde. Denn dadurch, dass ich Sonnenstrom in meinen Akku speichere, wird kein deutsches Kohlekraftwerk entlastet. Mit der Formel von ChatGPT wäre die Ersparnis aber - 2 kWh * Emissionsfaktor, also negativ (!) und somit falsch.

Vielleicht habe ich einen Denkfehler, aber das Ergebnis mit ChatGPT wäre das gleiche, wie bei deiner Rechung. PV Erzeugung 2 - Akkubeladung 2 = 0

> Fortführung des Beispiels, eine Stunde später:
> 
> Sonne scheint nicht mehr, ich schalte meine Verbraucher ein, Akku wird entladen:
> 
> Stromverbrauch: 2 kWh
> Akkuentladung: 2 kWh
> alles andere: 0 kWh

Da gebe ich dir Recht. Das passt tatsächlich nicht.

Vielleicht rechnen wir aber auch alle viel zu kompliziert. Im Prinzip geht es doch um die CO2 Ersparnis bei Reduzierung des Netzbezugs. Diesen Wert haben wir doch schon, nämlich die aktuelle Autarkie! Also einfach Autarkie * Emissionsfaktor = aktuelle Co2 Ersparnis. In dem Autarkiewert ist ja alles Enthalten (Netzbezug, PV Ertrag, Akku Be- und Entladung). 

Kleines Beispiel:
Annahme Emissionsfaktor = 0,5 kg CO₂/kWh, Hausverbrauch = 1 kW

1.) Autarkie 100% (also kein Netzbezug, ich spare 100% ggü. dem aktuellen Strommix): 100% * Emissionsfaktor * Hausverbrauch = Aktuelle CO2 Ersparnis

100% * 0,5 * 1 = 0,5 kg CO₂/kWh

2.) Autarkie 0% (also 100 % Netzbezug, ich spare 0% ggü. dem aktuellen Strommix: 0% * Emissionsfaktor * Hausverbrauch = Aktuelle CO2 Ersparnis

0% * 0,5 * 1 = 0 kg CO₂/kWh

3.) Autarkie 50% (also 50% Netzbezug, ich spare 50% ggü. dem aktuellen Strommix: 50% * Emissionskator * Hausverbrauch = Aktuelle CO2 Ersparnis

50% * 0,5 * 1 = 0,25 kg CO₂/kWh

Nun hat man die CO2 Ersparnis des eigenen Hausverbrauches im Gegensatz zum Netzbezug. Jetzt fehlt noch die Einspeisung ins Netz, welches ja nochmal zusätzliche CO2 Ersparnis erzeugt. Also müsste man noch zu der Rechnung die aktuelle Einspeisung in Netz * dem Emissionsfaktor in die Gleichung einsetzen.
Annahme Einspeisung 2 kW

Beispiel:
1.) (100% * 0,5 * 1) + (2 * 0,5) = 1,5 kg CO₂/kWh

Annahme Einspeisung 4 kW
1.) 100% * 0,5 * 1) + (4 * 0,5) = 2,5 kg CO₂/kWh

Das macht aus meiner Sicht am meisten sinn. Hoffe, dass ich keinen Denkfehler mache,…

**Comment by boecks** [comment]
Was auch immer das Ergebnis ist, ich speichere die Daten von Electricity Maps sicherheitshalber schon jetzt via MQTT Collector in InfluxDB. :-)

**Comment by ledermann** [comment]
@spirit0607 Sowohl in meiner Formel als auch in der von ChatGPT kommt die PV-Erzeugung gar nicht vor. Laut der ChatGPG-Formel ergäbe sich `0 - 0 + 0 - 2 + 0 = -2`, was nicht korrekt sein kann. Ja, man könnte auch mit der Autarkie rechnen, denn: `Autarkie = 100 - Netzbezug / Stromverbrauch`. Das könnte man umformen und in meine Formel einsetzen, was sie aber nicht einfacher macht.

@AlpenFlizzer In deiner Formel fehlt aber die Einspeisung. Die muss man meiner Ansicht nach auch berücksichtigen, weil die ja den Nachbarn dabei hilft, CO₂ einzusparen. Wenn ich in deinem Beispiel die Einspeisung mit 0 annehme (weil Netzbezug > 0), dann ergibt sich mit meiner Formel ebenfalls 10 g. 

Ich bin somit weiterhin der Ansicht, dass meine Formel korrekt ist :-)

```
CO₂-Reduktion = (Stromverbrauch - Netzbezug + Einspeisung) * Emissionsfaktor
```

Und ja, Emissionsfaktor schon mal einsammeln ist gut :)

**Reply by spirit0607** [reply]
Ja, du hast recht. Deine Formel stimmt. :-) Irgendwie hatte ich wohl Tomaten auf den Augen ;-) Aber die Umsetzung mit der Formel, dem dynamischen Emissionsfaktor und der dann ermittelten Co2 Ersparnis wäre für SOLECTRUS ein echter Mehrwehrt!

**Reply by spirit0607** [reply]
Aber nur mal noch so für mich zum Verständnis: 
Die aktuelle Formel in der Develop Version ist aber eine andere, oder? Hier wird aktuell einfach nur Stromerzeugung * 401 g/kWh gerechnet, oder?

**Reply by ledermann** [reply]
Ja, genau so ist es. Derzeitig ist der Faktor konstant `401 g/kWh`, es hat noch keine anderweitige Implementierung stattgefunden.

**Reply by spirit0607** [reply]
Okay, ja mit konstantem Faktor hatte ich mir schon gedacht. Den Collector zu bauen und zu implementieren ist ja vermutlich auch etwas aufwändiger. Ich hatte jetzt auch eher an die Hinterlegung der neuen Formel gedacht. Nämlich jetzt schonmal anstatt `PV-Erzeugung * Emissionsfaktor ` mit `CO₂-Reduktion = (Stromverbrauch - Netzbezug + Einspeisung) * Emissionsfaktor ` zu rechnen.

**Reply by ledermann** [reply]
Das Wesentliche an der neuen Berechnungsweise ist, dass sie **stündlich** erfolgen muss. Wenn man die neue Formel nur auf die **Summe** (eines Zeitraums) anwendet, ändert sich kaum etwas, denn:

```
Wegen Bilanzeigenschaft: Herkunft = Verwendung, also:
Netzbezug + Erzeugung + Akkuentnahme = Stromverbrauch + Einspeisung + Akkubeladung

Auflösen nach Erzeugung:
⇔ Erzeugung = Stromverbrauch - Netzbezug + Einspeisung + Akkubeladung - Akkuentnahme

und da Akkubeladung ≈ Akkuentnahme (bei größeren Zeiträumen):
⇔ Erzeugung ≈ Stromverbrauch - Netzbezug + Einspeisung

Somit gilt:

CO₂-Reduktion = (Stromverbrauch - Netzbezug + Einspeisung) * Emissionsfaktor
              ≈ Erzeugung * Emissionsfaktor
```

Die stündliche Berechnung ist allerdings mit InfluxDB nicht trivial, weil das zu großen Performance-Problemen führen kann. Hier muss ich mir noch etwas überlegen. 

Issue #3198 ist angelegt.

**Comment by Zeppelin-PV** [comment]
hallo Georg,
war ne Zeit lang verschollen und habe zu diesem Thema eine Frage
Was ist wenn mein Stromanbieter mir sagt dass er nur erneuerbare Energie in meine Steckdosen schickt
und mir hierzu dieses PDF vorhält
https://evf.de/dateien/privatkunden/strom/evf_strom_info_flyer_04-2021.pdf

habe das develop geladen und bin erstaunt was ich einspare obwohl ich Co² frei beliefert werde
Spare ich den Bundesdurchschnitt ein?

**Reply by ledermann** [reply]
Nicht mein Fachgebiet, aber ich verstehe das so:

Möglicherweise ist es tatsächlich so, dass dein Strom aus erneuerbaren Quellen kommt – aber das liegt keineswegs daran, dass du einen Vertrag mit EVF abgeschlossen hast. Dein Nachbar, der vermutlich einen anderen Anbieter gewählt hat, bekommt denselben Strom wie du – und der ist ein Mix aus unterschiedlichen Quellen.

Aber selbst wenn du dir 100%ig sicher bist, dass in deinem konkreten Fall der Strom grün ist, weil EVF vielleicht ein Kabel direkt vom Windrad zu deinem Haus gelegt hat, kann man es ja so sehen: Aufgrund deiner PV-Anlage verbrauchst du weniger externen Grünstrom, der somit anderen Verbrauchern zur Verfügung steht. Und irgendwo in der Kette wird dann ein deutsches Kohlekraftwerk etwas weniger verfeuern müssen.

Insofern: Ja, du sparst den Bundesdurchschnitt an CO₂ ein. Auf https://app.electricitymaps.com/zone/DE wird für Deutschland nur eine einzige Zone ausgewiesen. Das bedeutet, der CO₂-Ausstoß pro kWh wird bundesweit als identisch angenommen. Der Wert, den du dort siehst, ändert sich aber stündlich. 

Die Neuerung, die hier diskutiert wird, ist also: Statt mit konstant 401 g/kWh zu rechnen, soll stundenweise mit dem jeweils aktuellen Wert von electricitymaps.com gerechnet werden.

**Comment by spirit0607** [comment]
Das ist auch nicht mein genaues Fachgebiet, aber:

> ....,dass er nur erneuerbare Energie in meine Steckdosen schickt

Aber genau das ist der Grund, warum der Anteil an CO2 armer Stromproduktion in Deutschland steigt. Wenn alle privaten Haushalte und die ganze Industrie nur Strom aus Windanlagen, PV, o.ä. beziehen würden, würde der Anteil immer weiter gen 100% steigen und somit wäre der Strommix in Deutschland klimaneutral.
 
Als Ergebnis kann man meiner Meinung nach nur seine eigene PV Anlage mit dem deutschen Strommix vergleichen, da dies nunmal die Referenz ist. Wenn man das mit seinem eigenen Stromvertrag vergleicht, würde das dazu führen, dass man nie eine Ersparnis hat, da man ggü Null CO2 nichts einsparen kann. Ich habe sogar mal von einer Firma gehört, die sich aus Nachhaltigkeitsgründen deshalb gegen eine PV Anlage entschieden hat. Argumentation war hier, dass bereits 100% Strom aus klimaneutraler Wasser,- und Windkraft bezogen wird. Beim Bau und Installation wird CO2 freigesetzt, welches dazu führt, dass die eigene Klimabilanz der Firma negativ wird. Wenn man die Fakten so auslegt, kann das mit der Energiewende nie etwas werden.

Eine PV Anlage hilft, Strom CO2 neutraler zu produzieren und somit CO2 einzusparen und genau dies wird mit der Berechnung meines Erachtens korrekt dargestellt.

**Comment by Zeppelin-PV** [comment]
... noch eine Frage zum CO² Thema
unsere große PV-Anlage wird mit Solectrus visualisiert,
das BKW das nebenher läuft mit HomeAssistant (Energie)
dort kann ich relativ einfach die electricity.map integrieren
![grafik](https://github.com/solectrus/solectrus/assets/126392805/3dd8869f-0bcc-4103-b3d5-7785b24cc512)
wäre es machbar diesen Wert rüber zu holen zu Solectrus, dann wäre er ja immer aktuell
die Konstellation wäre hier von einem Raspberry (HomeAssistant) auf den Raspberry (Solectrus)

**Reply by ledermann** [reply]
Mit dem MQTT-Collector könntest du die CO₂-Intensität vermutlich dort abholen. Aber es wird einen neuen Collector geben, weil ich hierfür eine Abhängigkeit vom Home Assistant vermeiden möchte. Den neuen Collector zu benutzen, wird der einfachere Weg sein.
